### PR TITLE
Proposal: simpler LayerStack and DerivedLayers

### PR DESCRIPTION
**Pull request by simbilod**
The current way to deal with "derived layers" (i.e. physical levels on the chip that do not map 1:1 to GDS design layers) is very convoluted. It requires defining etch and grow levels, with flags for what is etched into what, which borders on process emulation. This also does not generalize well to other common mapping between GDS layers and physical elements on the chip, for instance marker layers modifying shared process steps.

I propose instead, at the gdsfactory level, to only consider the final fabricated levels of the chip. These layers should be easily defined from **arbitrary combinations of GDS layers**. An example could look something like this:

```
    layer1 = GDSLayer(layer=(1, 0))
    layer2 = GDSLayer(layer=(2, 0))

    ls = LayerStack(
        layers={
            "layerlevel_layer1": LayerLevel(layer=layer1, thickness=10, zmin=0),
            "layerlevel_layer2": LayerLevel(layer=layer2, thickness=10, zmin=10),
            "layerlevel_and_layer": LayerLevel(
                layer=layer1 & layer2,
                thickness=10,
                zmin=0,
                derived_layer=GDSLayer(layer=(3, 0)),
            ),
            "layerlevel_xor_layer": LayerLevel(
                layer=layer1 ^ layer2,
                thickness=10,
                zmin=0,
                derived_layer=GDSLayer(layer=(4, 0)),
            ),
            "layerlevel_not_layer": LayerLevel(
                layer=layer1 - layer2,
                thickness=10,
                zmin=0,
                derived_layer=GDSLayer(layer=(5, 0)),
            ),
            "layerlevel_or_layer": LayerLevel(
                layer=layer1 | layer2,
                thickness=10,
                zmin=0,
                derived_layer=GDSLayer(layer=(6, 0)),
            ),
            "layerlevel_composed_layer": LayerLevel(
                layer=layer1 - (layer1 ^ layer2),
                thickness=10,
                zmin=0,
                derived_layer=GDSLayer(layer=(7, 0)),
            ),
        }
    )

    # Test with simple component
    import gdsfactory as gf

    c = gf.Component()

    rect1 = c << gf.components.rectangle(size=(10, 10), layer=(1, 0))
    rect2 = c << gf.components.rectangle(size=(10, 10), layer=(2, 0))
    rect2.dmove((5, 5))
    c.show()
```
image

Now get the component with derived layers according to the stack above:
```
    c = get_component_with_derived_layers(c, ls)
    c.show()
```

(3,0) is the intersection:

image

(4,0) is the exclusive or:

image

(5,0) is a substraction:

image

(6,0) is union:

image

We can also compose the operations; (7,0) substracts the X0R from layer 1:

image

We can revisit process emulation and/or simulation in gplugins.

If you think this is a good idea I can make tests and fix visualization scripts

### CODE REVIEW

**Comment by simbilod** [comment]
I'm also removing all the clutter from as discussed in #2556

**Comment by joamatab** [comment]
looks great, notice there are some conflicts

File: gdsfactory/technology/layer_stack.py [hunk 1]
```diff
@@ -8,65 +8,212 @@
...
 from rich.table import Table
 
-from gdsfactory.component import Component, boolean_operations
+from gdsfactory.component import Component
 
 if TYPE_CHECKING:
     from gdsfactory.technology import LayerViews
 
 
+class GDSLayer(BaseModel):
```
Comment (joamatab) [inline gdsfactory/technology/layer_stack.py:17]: How about calling it LogicalLayer?

**Review by joamatab — COMMENTED** [review]
looks great!

File: gdsfactory/technology/layer_stack.py (hunk 1 above)
Comment (simbilod) [reply gdsfactory/technology/layer_stack.py]: done!

**Comment by simbilod** [comment]
get_klayout_3d_script needs fixing, but it should be easy to convert the boolean operations in the layerstack to klayout script booleans

**Comment by simbilod** [comment]
there's issues with Pydantic to_json, will check again later

**Comment by simbilod** [comment]
> there's issues with Pydantic to_json, will check again later

the pydantic serialization error is fixable by allowing layers to be int, is this acceptable behavior @simbilod @sebastian-goeldi ?

File: gdsfactory/technology/layer_stack.py [hunk 2]
```diff
@@ -14,59 +14,234 @@
     from gdsfactory.technology import LayerViews
 
 
+class LogicalLayer(BaseModel):
+    """GDS design layer."""
+
+    layer: tuple[int, int] | kf.LayerEnum | int
```
Comment (simbilod) [inline gdsfactory/technology/layer_stack.py:20]: here

File: gdsfactory/technology/layer_stack.py [hunk 3]
```diff
@@ -14,59 +14,234 @@
...
+    Overloads operators for simpler expressions.
+
+    Attributes:
+        input_layer1: primary layer comprising the derived layer. Can be a GDS design layer (kf.LayerEnum, tuple[int, int]), or another derived layer.
+        input_layer2: secondary layer comprising the derived layer. Can be a GDS design layer (kf.LayerEnum, tuple[int, int]), or another derived layer.
+        operation: operation to perform between layer1 and layer2. One of "and", "or", "xor", or "not" or associated symbols.
+    """
+
+    layer1: LogicalLayer | DerivedLayer | int
+    layer2: LogicalLayer | DerivedLayer | int
```
Comment (simbilod) [inline gdsfactory/technology/layer_stack.py:122]: and here

**Comment by sebastian-goeldi** [comment]
> > there's issues with Pydantic to_json, will check again later
> 
> the pydantic serialization error is fixable by allowing layers to be int, is this acceptable behavior @joamatab @sebastian-goeldi ?

The problem with `int` is, it's not a good identifier, even `LayerEnum` is questionable in this case. Anything that should live with layers outside or directly tied info, should use something more akin to `gf.kdb.LayerInfo`, so tuple with layer datatype or a string with layer name are better choices.

An int should be stricly used for layout purposes (it's even not so good to pass it through function calls tbh as they will be serialized as ints)

To get tuple from LayerEnum:
`tuple(LAYER.WG)` for example

To get name:
`LAYER.WG.name`

To get layerenum from tuple:
`LAYER(gf.kcl.layer(*layer_tuple))` (or `layer_tuple[0],layer_tuple[1]` instead o `*...` of course.

Get LayerEnum by name:
`LAYER[layer_name]`

**Comment by joamatab** [comment]
when will this PR be ready to review? we plan to release gdsfactory8 in the coming days

**Comment by simbilod** [comment]
Of we can disable export KLayout 3D view, it can be done soon. Otherwise it will take longer

**Comment by joamatab** [comment]
yes, we can fix the export to klayout3D later

**Comment by simbilod** [comment]
I've implemented @HelgeGehring 's suggestions, it still works exactly as I showed above

@joamatab this is ready to review. I've added a NotImplementedError on the klayout 3d script for now. The only sketchy thing is the int type allowed for layer object to fix pydantic issues. Also the commit history is a bit messy, we can rebase against main if you want.

**Comment by simbilod** [comment]
Some export tests are still failing, probably just need to call `layer.layer` in some places. I can look tomorrow

**Comment by joamatab** [comment]
Sounds good, 

will this work with the plugins? like gmsh?

im having lots of issues with the gdsfactory8 brach in gplugins

i think it would be great to use klayout instead of shapely there if possible

**Comment by simbilod** [comment]
> Sounds good,
> 
> will this work with the plugins? like gmsh?
> 
> im having lots of issues with the gdsfactory8 brach in gplugins
> 
> i think it would be great to use klayout instead of shapely there if possible



> Sounds good,
> 
> will this work with the plugins? like gmsh?
> 
> im having lots of issues with the gdsfactory8 brach in gplugins
> 
> i think it would be great to use klayout instead of shapely there if possible

The meshing should mostly be fine, it only relies on get_component or get_component_with_derived_layers, so if those work it will be OK. We might need to add a layer.layer here and there at most.

**Comment by sebastian-goeldi** [comment]
I think for layers in general, I think we should just pass them as strings (i.e. layer name).

The longer the more I think it's easier to do it that way. Getting them can be as easy as `kcl.get_layer`. This also allows for easier and more meaningful (de)serialization of layers as the index itself has next to no meaning (except during construction)

**Pull request merged**